### PR TITLE
Custom properties vip:dot

### PR DESCRIPTION
**Pull request by sandepat**
# Changes in VIP-portal regarding custom properties

## Custom dot inputs
- Parse custom properties of Boutiques descriptor and read the dot inputs

## Handling of Command Line Flag Inputs
- Implemented a feature to detect command-line-flag inputs included in the Boutiques custom properties as dot inputs.
- Display a warning message at application import time, ensuring better user awareness and error prevention.

### CODE REVIEW

**Review by axlbonnet — APPROVED** [review]
Fine to me.
@camarasu do you want to review ?

**Comment by camarasu** [comment]
Fine for me, too, thank you. 
Waiting for the last updates discussed earlier (handling of results-directory and checking of dot inputs) before merging the PR.

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/bean/boutiquesTools/BoutiquesApplication.java [hunk 1]
```diff
@@ -38,6 +41,8 @@ public class BoutiquesApplication implements IsSerializable {
     private Set<BoutiquesOutputFile> outputFiles = new HashSet<>();
     private Map<String, String> tags = new HashMap<>();
     private String jsonFile;
+    private Set<String> vipDotInputIds;
+    private String resultDirs;
```
Comment (axlbonnet) [inline vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/bean/boutiquesTools/BoutiquesApplication.java:45]: rename field and getter/setter with `vipDotIncludesResultsDir`, and change type to boolean.

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java [hunk 2]
```diff
@@ -114,8 +114,9 @@ public BoutiquesApplication parseApplication(String descriptor) throws InvalidBo
         // Custom property
         JSONObject customObject = getObjectValue(parsedDescriptor, "custom", true);
         if (customObject != null) {
-            String vipImagePath = getStringValue(customObject, "vip:imagepath", true);
-            application.setVipContainer(vipImagePath);
+            application.setVipContainer(getStringValue(customObject, "vip:imagepath", true));
+            application.setVipDotInputIds(getArrayValueAsStringSet(customObject, "vip:dot", true));
+            application.setVipDotResultDirs(getStringValue(customObject, "vip:dot-result-directory", true));
```
Comment (axlbonnet) [inline vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java:119]: Rename to `vip:dot-with-results-directory`.
Adapt to `vip:dot-result-directory` being a boolean.

File: vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java [hunk 3]
```diff
@@ -158,6 +163,35 @@ private static void verifyBoutiquesTool(BoutiquesApplication boutiquesTool)
...
+        }
+
+        // Extract the IDs from inputs
+        Set<String> inputIds = application.getInputs().stream().map(BoutiquesInput::getId).collect(Collectors.toSet());
+        // Check if all vipDotInputIds are in inputs
+        if (!inputIds.containsAll(vipDotInputIds)) {
+            Set<String> incorrectInputs = new HashSet<>(vipDotInputIds);
+            incorrectInputs.removeAll(inputIds);
+            String warningMessage = "<b>" + String.join(", ", incorrectInputs) + "</b> appears in vipDotInputIds but not in inputs. Please ensure all ids are correct.";
+            Layout.getInstance().setWarningMessage(warningMessage);
```
Comment (axlbonnet) [inline vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java:193]: in this case, it should be a error with an exception thrown.

File: vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java [hunk 4]
```diff
@@ -158,6 +163,35 @@ private static void verifyBoutiquesTool(BoutiquesApplication boutiquesTool)
...
+        
+        commonValues.retainAll(commandLineFlags);
+
+        if (!commonValues.isEmpty()) {
+            String warningMessage = "<b>" + String.join(", ", commonValues) + "</b> appears as command-line flag input(s), it should not be included in Dot iteration. Importing it may cause functionality issues, although the application will still be imported.";
+            Layout.getInstance().setWarningMessage(warningMessage);
+        }
+
+        // Extract the IDs from inputs
+        Set<String> inputIds = application.getInputs().stream().map(BoutiquesInput::getId).collect(Collectors.toSet());
```
Comment (axlbonnet) [inline vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java:187]: You may add this as a BoutiquesApplication method. (optionally)

File: vip-portal/src/main/resources/vm/gwendia-standalone.vm [hunk 5]
```diff
@@ -5,78 +5,99 @@
...
-  #set($type="string")
-  #if($input.getType().getCamelName()=="File")
-    #set($type="URI")
-  #end
-  #if($input.getType().getCamelName()=="Flag")
-    #if($input.getDefaultValue() && $input.getDefaultValue()!="")
-      <source name="$input.getId()" type="string" optional="$input.isOptional()" default="$input.getDefaultValue()" vip-type-restriction="flag" pretty-name="$input.getName()">
-    #else
-      <source name="$input.getId()" type="string" optional="$input.isOptional()" default="false" vip-type-restriction="flag" pretty-name="$input.getName()">
+    #foreach( $input in $tool.getInputs() )
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/gwendia-standalone.vm:8]: I thnk VS or something else changed the indentation of velocity code and it generetad a lot of diff.
Could it be possible to keep the original indentation ?

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java (hunk 2 above)
Comment (sandepat) [reply vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java:119]: change to vip:dot-with-results-directory and adapt to the method getBooleanValue

File: vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java (hunk 4 above)
Comment (sandepat) [reply vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java:187]: Added the getinputIds method to BoutiquesApplication

File: vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java (hunk 3 above)
Comment (sandepat) [reply vip-application-importer/src/main/java/fr/insalyon/creatis/vip/applicationimporter/client/view/applicationdisplay/DisplayTab.java:193]: throwing ApplicationImporterException

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/bean/boutiquesTools/BoutiquesApplication.java (hunk 1 above)
Comment (sandepat) [reply vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/bean/boutiquesTools/BoutiquesApplication.java:45]: renamed to vipDotIncludesResultsDir

File: vip-portal/src/main/resources/vm/gwendia-standalone.vm (hunk 5 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/gwendia-standalone.vm:8]: indeed, restore original indentation.

**Pull request merged**